File: DS-n-Algos-Mega-Archive/JAVASCRIPT/Data.Structures.and.Algorithm/JavaScript/Linked-Lists/singly-linked-list.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  constructor(head) {
    this.head = null;
  }

  displayList() {
    let tempPtr = this.head;
    let output = '';
    while (tempPtr !== null) {
      output += `${tempPtr.data} -> `;
      tempPtr = tempPtr.next;
    }
    output += 'Null';
    console.log(output);
  }

  addAtBeginning(data) {
    let newNode = new Node(data);
    newNode.next = this.head;
    this.head = newNode;
  }

  addAtEnd(data) {
    let newNode = new Node(data);
    let tempPtr = this.head;
    while (tempPtr.next) {
      tempPtr = tempPtr.next;
    }

    tempPtr.next = newNode;
  }

  addAtNthPosition(data, position) {
    let tempPtr = this.head;
    for (let i = 0; i < position - 1; i++) {
      tempPtr = tempPtr.next;
    }

    let newNode = new Node(data);
    newNode.next = tempPtr.next;
    tempPtr.next = newNode;
  }

  removeFromBeginning() {
    if (this.head.next) this.head = this.head.next;
    else console.log('Items unavailable');
  }

  removeFromEnd() {
    let tempPtr = this.head;
    while (tempPtr.next.next) {
      tempPtr = tempPtr.next;
    }

    tempPtr.next = null;
  }

  removeFromNthPosition(position) {
    let tempPtr = this.head;
    for (let i = 0; i < position - 1; i++) {
      tempPtr = tempPtr.next;
    }

    tempPtr.next = tempPtr.next.next;
  }
}

let list = new LinkedList();
list.addAtBeginning(10);
list.addAtBeginning(20);
list.addAtBeginning(30);
list.addAtNthPosition(200, 2);
list.addAtEnd(0);
list.addAtEnd(100);
list.displayList();

list.removeFromBeginning();
list.displayList();

list.removeFromEnd();
list.displayList();

list.removeFromNthPosition(2);
list.displayList();

File: DS-n-Algos-Mega-Archive/JAVASCRIPT/Data.Structures.and.Algorithm/Kotlin/src/linked_lists/singly-linked-list.kt
package linked_lists

data class Node(
    var data: Int,
    var next: Node? = null
)

data class LinkedList(var head: Node? = null) {
    fun displayList() {
        var tempPtr = head
        var output = ""
        while (tempPtr != null) {
            output += "${tempPtr.data} -> "
            tempPtr = tempPtr.next
        }
        output += "Null"
        println(output)
    }

    fun addAtBeginning(data: Int) {
        val newNode = Node(data)
        newNode.next = head
        head = newNode
    }

    fun addAtEnd(data: Int) {
        val newNode = Node(data)
        var tempPtr = head
        while (tempPtr!!.next != null) {
            tempPtr = tempPtr.next
        }
        tempPtr.next = newNode
    }

    fun addAtNthPosition(data: Int, position: Int) {
        var tempPtr = head
        for (i in 1 until position) {
            tempPtr = tempPtr!!.next
        }
        val newNode = Node(data)
        newNode.next = tempPtr!!.next
        tempPtr.next = newNode
    }

    fun removeFromBeginning() {
        if (head!!.next != null) head = head!!.next
        else println("Items unavailable")
    }

    fun removeFromEnd() {
        var tempPtr = head
        while (tempPtr!!.next!!.next != null) {
            tempPtr = tempPtr.next
        }
        tempPtr.next = null
    }

    fun removeFromNthPosition(position: Int) {
        var tempPtr = head
        for (i in 1 until position) {
            tempPtr = tempPtr!!.next
        }
        tempPtr!!.next = tempPtr.next!!.next
    }
}

fun main() {
    val list = LinkedList()
    list.addAtBeginning(10)
    list.addAtBeginning(20)
    list.addAtBeginning(30)
    list.addAtNthPosition(200, 2)
    list.addAtEnd(0)
    list.addAtEnd(100)
    list.displayList()
    list.removeFromBeginning()
    list.displayList()
    list.removeFromEnd()
    list.displayList()
    list.removeFromNthPosition(2)
    list.displayList()
}